### ISSUE TITLE
please add an option to enable debug symbols only

### ISSUE THREAD

**Issue by vicuna** [ocaml/ocaml#6238]
**Original bug ID:** 6238
**Reporter:** @ygrek
**Assigned to:** @mshinwell
**Status:** acknowledged (set by @mshinwell on 2013-11-18T16:29:38Z)
**Resolution:** open
**Priority:** normal
**Severity:** feature
**Target version:** undecided
**Category:** back end (clambda to assembly)
**Related to:** #6728
**Monitored by:** @gasche @ygrek @Yakobowski @alainfrisch

## Bug description

Rationale:
currently option -g mixes several things up - generation debug symbols, storing backtrace information, adding more precise bounds checking, disabling some optimizations (in bytecode).
For profiling one usually wants the same binary code that is run in production but with debug symbols. So the speed of the code is the same, but there is some extra information available on demand, and this build can be run on production at highest speed and then profiled live if any problem arises. Option -g makes this approach less attractive because of incurred performance cost (e.g. storing backtrace).
Can we please have another option to _only_ write out (native, i.e. DWARF) debugging info and perform no other changes to generated code?

**Comment by vicuna** [comment]
**Comment author:** @mshinwell

I think this is a reasonable thing to do.  Watch this space.

**Comment by vicuna** [comment]
**Comment author:** @gasche

While I think this is a reasonable feature, I think that in the long term the way to go is rather to provide more control to performance-conscious users of which raises will trigger backtrace recording (eg. the raise-variants work of Alain). Indeed, there should be a sweet spot where most exceptions used for control flow do not record any trace, only actually exceptional exceptions do, and you have a system that is both efficient (essentially as the no-trace one) and auditable in case of unplanned failure (a good reason to have *some* backtraces even in production). I can understand people disabling everything for absolute performances (and right now it's probably easier to implement), but long-term I think that's more of a minority use-case.

Of course if someone is ready to do the work, that's great in any case.

**Comment by vicuna** [comment]
**Comment author:** @damiendoligez

Gabriel, your argument addresses backtraces only, not the bound-checking and disabled optimizations.

**Comment by vicuna** [comment]
**Comment author:** @alainfrisch

> more precise bounds checking

Have you seen cases where keeping multiple call sites for caml_ml_array_bound_error yields a noticeable performance penalty?

> disabling some optimizations (in bytecode)

Since the request is to have a mode with maximal runtime performance (while keeping debugging information), I don't think that bytecode is relevant.

Alain

**Comment by vicuna** [comment]
**Comment author:** @damiendoligez

In any case, the request is for adding debugging info to the executable, while guaranteeing no run-time overhead, and this seems quite reasonable to me.

**Comment by vicuna** [comment]
**Comment author:** @gasche

Mark, is it time for a gentle ping?

**Comment by vicuna** [comment]
**Comment author:** @mshinwell

Please see the comment on the pull request

https://github.com/ocaml/ocaml/pull/574

**Comment by jberdine** [comment]
Just for info, we would appreciate such a full-speed plus debug info binary mode.

**Comment by gasche** [comment]
Thanks @jberdine for your help with bug triaging (and your feedback) these days.

**Comment by damiendoligez** [comment]
@gasche do you think this could be a "good first issue"?

**Comment by gasche** [comment]
Yes, but I think that we should start gathering a consensus on the naming of such an option, because while the implementation may be relatively simple, picking a good name probably is not.

**Comment by Octachron** [comment]
The feature wish seems reasonable to me, and seems to have only be stuck due to naming paralysis.

**Comment by cjc25** [comment]
In case someone is waiting for name inspiration, `clang` uses `-gmlt` ("minimum line table") to include symbol info only (e.g. function {names, filenames, line numbers}). `gcc` analogously has `-g1`

**Comment by ygrek** [comment]
`-glevel` options for gcc only change the amount of debugging information emitted, it doesn't change emitted code, this is what this issue is about

**Comment by nojb** [comment]
Sorry for the naïve question, but as the issue description above mentions native code, what "debug symbols" are we talking about? The location information for each instruction?

**Comment by ygrek** [comment]
> (native, i.e. DWARF) debugging info

previously (at the time this issue was created) it was function names, afaihs now we also have source locations
